Data/ByteString/Builder/Prim.hs

```diff
@@ -672,50 +673,87 @@ primMapLazyByteStringBounded w =
 --
 -- @since 0.11.0.0
 cstring :: Addr# -> Builder
-cstring =
-    \addr0 -> builder $ step addr0
-  where
-    step :: Addr# -> BuildStep r -> BuildStep r
-    step !addr !k br@(BufferRange op0@(Ptr op0#) ope)
-      | W8# ch == 0 = k br
-      | op0 == ope =
-          return $ bufferFull 1 op0 (step addr k)
-      | otherwise = do
-          IO $ \s -> case writeWord8OffAddr# op0# 0# ch s of
-                       s' -> (# s', () #)
-          let br' = BufferRange (op0 `plusPtr` 1) ope
-          step (addr `plusAddr#` 1#) k br'
-      where
-        !ch = indexWord8OffAddr# addr 0#
+cstring = \addr0 -> builder $ \k br -> do
+#if __GLASGOW_HASKELL__ >= 811
+    let len = cstringLength# addr0
+#else
+    (I# len) <- fromIntegral <$> S.c_strlen (Ptr addr0)
+#endif
+    cstring_step addr0 len k br
+{-# INLINE cstring #-}
+
+cstring_step :: Addr# -> Int# -> BuildStep r -> BuildStep r
+cstring_step !addr !len !k br@(BufferRange op0 ope)
```

Comment:
It would make sense for this to just wrap `byteString` or `byteStringCopy` from `Builder.Internal`.

Reply:
I think you mean `wrappedBytesCopyStep`, which is the closest to usable, because it does not need an input `ByteString` (foreign pointer, a...), but works directly with an input start/end pointer pair (Buffer range).  This is not currently exported from `Data/ByteString/Builder/Internal.hs`, but perhaps we could move the `cstring` code there?
The new implementation has nothing to do with the `BoundedPrim` API...

Reply:
Done, roughly as suggested. This seems to give a small additional performance boost.

Reply:
Well, a `ByteString` is just an `unsafePackLenLiteral` away, but moving the code is just as well. The original `cstring` code also had nothing to do with the (very public) `BoundedPrim` API. I have no idea why it was ever put there to begin with...